metis/lib/server/controllers/download_controller.rb

```diff
@@ -11,10 +14,34 @@ def download
 
     raise Etna::Error.new('File not found', 404) unless file && file.has_data?
 
+    if @params.fetch(:thumbnail, nil)
+      thumbnail = Vips::Image.thumbnail(file.data_block.location, 240)
+      mimetype = MimeMagic.by_path(file.file_name)
+
+      case mimetype.to_s
+      when /jpe?g$/
+        buffer = thumbnail.jpegsave_buffer
```

Comment:
Where do these buffers get saved btw?  Is the buffer consistent per path or unique on each call?

Reply:
Or, is the thumbnail entirely in memory?

Reply:
It looks like Rack streams the object (see line 34 below), so it's all in memory. This leads me to wonder how much memory this uses and how efficient the process is, as some of these TIFF and JPEG files can get quite large (if they are showing an H&E stain of a tumor at high magnification e.g.)

Reply:
It's entirely in memory. I'm not sure how to measure (tips?), but I manually tested with a 1GB TIFF file, and it streamed back really quickly, and I would say it's not obvious that it's generating the thumbnail on the fly. vips is more performant than ImageMagick (you can check [their benchmarks](https://github.com/libvips/libvips/wiki/Speed-and-memory-use)).

You can test it on timur-test-1.ucsf.edu, just upload a giant image to metis-test-1.ucsf.edu...

Reply:
Of course, testing it locally, it's highly dependent on my laptop's specs. Not sure how those compare to the VMs.

Reply:
The transformation may itself be streamable, which means it wouldn't require much memory.  It would just read chunks from disk, do the transform for that chunk, and pass it over the wire before continuing.

That said, it _could_ be CPU intensive if this is happening for many thumbnails without any caching.  Is it possible for us to atleast set an ETag and observe If-None-Match to ensure clients atleast can attempt to cache the resource?
We could use the basename of the chunk (it's computed md5) as the unique identifier.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag

Reply:
Oh you know what, it isn't streaming _from the source to the network_.  The buffer contains the full transform, after which it streams the buffered response.  That means that it probably _can_ be CPU intensive, especially since it does the work eagerly.

Reply:
Of note: the search page on timur will render thumbnails, so this might trigger 200 x number-of-image-columns simultaneous thumbnail requests.

Reply:
Oh, I'll test that...200 would be a lot to process on the fly.